c10/core/SymbolicShapeMeta.cpp

```diff
@@ -114,57 +130,188 @@ static bool definitely_true(const SymBool& b) {
 }
 
 SymBool SymbolicShapeMeta::compute_is_non_overlapping_and_dense_dim4() const {
-  if (definitely_true(is_contiguous_)) {
+  init_is_contiguous();
+  if (definitely_true(is_contiguous())) {
     return true;
   }
-  if (definitely_true(is_channels_last_contiguous_)) {
+  init_is_channels_last_contiguous();
+  if (definitely_true(is_channels_last_contiguous())) {
     return true;
   }
-  return is_contiguous_ | is_channels_last_contiguous_ |
+  return is_contiguous() | is_channels_last_contiguous() |
       compute_non_overlapping_and_dense();
 }
 
 SymBool SymbolicShapeMeta::compute_channels_last_contiguous_3d_dim5() const {
-  if (definitely_true(is_channels_last_contiguous_)) {
+  init_is_channels_last_contiguous();
+  if (definitely_true(is_channels_last_contiguous())) {
     return false;
   }
-  return ~is_channels_last_contiguous_ & compute_channels_last_contiguous_3d();
+  return ~is_channels_last_contiguous() & compute_channels_last_contiguous_3d();
 }
 
 SymBool SymbolicShapeMeta::compute_channels_last_2d_dim5() const {
-  if (definitely_true(is_channels_last_3d_contiguous_)) {
+  init_is_channels_last_3d_contiguous();
+  if (definitely_true(is_channels_last_3d_contiguous())) {
     return false;
   }
-  return ~is_channels_last_3d_contiguous_ &
+  return ~is_channels_last_3d_contiguous() &
       compute_strides_like_channels_last_2d();
 }
 
 SymBool SymbolicShapeMeta::compute_channels_last_3d_dim5() const {
-  if (definitely_true(is_channels_last_)) {
+  if (definitely_true(is_channels_last())) {
     return false;
   }
-  return ~is_channels_last_ & compute_strides_like_channels_last_3d();
+  return ~is_channels_last() & compute_strides_like_channels_last_3d();
 }
 
 SymBool SymbolicShapeMeta::compute_is_non_overlapping_and_dense_dim5() const {
-  if (definitely_true(is_contiguous_)) {
+  if (definitely_true(is_contiguous())) {
     return true;
   }
-  if (definitely_true(is_channels_last_contiguous_)) {
+  if (definitely_true(is_channels_last_contiguous())) {
     return true;
   }
-  if (definitely_true(is_channels_last_3d_contiguous_)) {
+  if (definitely_true(is_channels_last_3d_contiguous())) {
     return true;
   }
-  return is_contiguous_ | is_channels_last_contiguous_ |
-      is_channels_last_3d_contiguous_ | compute_non_overlapping_and_dense();
+  return is_contiguous() | is_channels_last_contiguous() |
+      is_channels_last_3d_contiguous() | compute_non_overlapping_and_dense();
 }
 
 SymBool SymbolicShapeMeta::compute_is_non_overlapping_and_dense_anydim() const {
-  if (definitely_true(is_contiguous_)) {
+  if (definitely_true(is_contiguous())) {
     return true;
   }
-  return is_contiguous_ | compute_non_overlapping_and_dense();
+  return is_contiguous() | compute_non_overlapping_and_dense();
+}
+
+void SymbolicShapeMeta::set_numel(SymInt val) const {
+  std::scoped_lock lock{mutables_};
+  if (has_numel()) {
+    return;
+  }
+  numel_ = std::move(val);
+  available_.fetch_or(numel_avail);
```

Comment:
Okay, here is a thread safety strategy. I use double-checked locking where the available bit in an atomic int indicates the variable is ready to read, and the write to the variable is protected by a mutex. The actual values are computed outside of the mutex lock which creates a small risk of calculating it twice, but the SymInt calculations cannot be done under the mutex since it might call arbitrary python code.

Reply:
Recompute is fine, nice trick.